### ISSUE TITLE
rename min_version to max_launch_stage

### ISSUE THREAD

**Issue by melinath** [hashicorp/terraform-provider-google#17080]
# What kind of contribution is this issue about?

* MMv1-based resource

# Details

min_version is a bit of a confusing name, at least to me. I would expect that in an ordering of alpha -> beta -> GA, `min_version: beta` would mean that a feature is only in beta and GA - but actually it is only in alpha and beta. It would be nice to rename to max_launch_stage (or at least max_version), which has a more intuitive meaning.

**Comment by melinath** [comment]
Maybe max_channel to avoid confusion with v1, v2, etc?

**Comment by c2thorn** [comment]
What about just "launch_stage"?
The majority of uses will just be "launch_stage: beta"

**Comment by melinath** [comment]
SGTM

**Comment by zli82016** [comment]
We should also renaming `min_version` in EAP provider.

**Comment by melinath** [comment]
Summarizing discussion: The problem with launch_stage it that it's a term that usually refers to API feature launch stages, which don't necessarily correspond to the provider versions that the features are supported in. Possibly `provider_version` would be better? Or maybe some other option.

**Comment by rileykarson** [comment]
I suspect we want to use terminology like "release channels" long-term if/when we do change. The cost to change is pretty high and we may not gain significant clarity with anything though.

> I would expect that in an ordering of alpha -> beta -> GA, min_version: beta would mean that a feature is only in beta and GA - but actually it is only in alpha and beta. It would be nice to rename to max_launch_stage (or at least max_version), which has a more intuitive meaning.

I have the opposite intuition FWIW. The `min_version` is the minimum [provider] version where a feature is available. (Keeping in mind that "version" typically refers to our version channels, and "release(s)" to semver numbered releases)

**Comment by melinath** [comment]
`min` generally indicates that something is present in everything "higher" than something. In the ordering of our providers, alpha -> beta -> GA, GA is intuitively "higher" than beta, because it comes later in the release of features. So `min_version: beta` would intuitively mean that a feature is in beta and GA. However, it actually means that the feature is in alpha and beta.

I'm guessing that you're intuitively sorting GA as "lower" than beta? That's surprising to me, since this seems pretty clear. But I guess it's not! Perhaps we need to collect evidence (at a minimum surveying our team, maybe also some off-team contributors) to see how confusing people find this.

**Comment by zli82016** [comment]
`min_version: beta` is confusing for me. I thought it means that the feature is in beta, ga. `max_version: beta` or similar is easier for me to understand.  


In the [code](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/api/product/version.go#L22), `ga` is lower than `beta`. 
`var ORDER = []string{"ga", "beta", "alpha", "private"}` 
Maybe that is the reason to use `min_version`?

**Comment by melinath** [comment]
> Maybe that is the reason to use min_version?

Yeah, that's right. I assume that ordering made sense to the person who originally wrote that code, but I've always found it confusing.

**Comment by rileykarson** [comment]
> min generally indicates that something is present in everything "higher" than something. In the ordering of our providers, alpha -> beta -> GA, GA is intuitively "higher" than beta, because it comes later in the release of features. So min_version: beta would intuitively mean that a feature is in beta and GA. However, it actually means that the feature is in alpha and beta.

I would caution that intuition is different for different folks. My model for min/max is that the versions are supersets of one another, i.e. alpha > beta > GA, and GA is the smallest / minimum. A minimum version for something is the lowest version in that hierarchy where it is available.

And regardless of personal intuition, there's the code @zli82016 referenced that provides a de facto definition [and has for 7+ years](https://github.com/GoogleCloudPlatform/magic-modules/pull/243).

---

This is not a perfect model by any means, and there may not be one- there are many different kinds of versions / stability levels / release channels in GCP that we abstract over and will likely be more in the future (we've already seen significant definition drift on `google-beta` as GCP moved from channel-based API versioning to visibility-based). I don't believe there is strong consensus on ordinal ordering for release channels / stability levels, such as an independent spec / RFC.

Moving away from the current model has a churn cost. We need to ensure that any change we make is both justified in itself (i.e. being clearer to a general audience, ensuring we're not just flipping based on individual stylistic preference) and that the cost of churn is justified- open PRs, muscle memory/current knowlege, outdating old question responses, etc.

**Comment by melinath** [comment]
yes, it sounds like we agree that getting evidence so that we're not relying on personal intuition seems reasonable.

**Comment by rileykarson** [comment]
> We do also need to keep in mind that most contributors will be first-time or infrequent contributors, so the benefit to new contributors may quickly outweigh the cost to existing contributors.

This is a good consideration, yes. I did- that's why I looked for external specs/rfcs/stability or release channels for other projects and brought up churn cost.

Please keep absolute impact in mind. What I'd consider the steel thread for this issue is:

- The amended size on this ticket suggests it will take 16h or so directly.
- After making the change, existing maintainers and contributors will need to adjust. That likely doubles the cost, adding another 16h.
- New or returning contributors will save some time. Over a 2y horizon there'll likely be 250 or so of them. The provider versions are not very complicated- they can be communicated in 2-3 paragraphs today, taking at most 10m/contributor. The cost floor of an improved, more intuitive model is likely 5m/contributor as they're not coming in with external knowledge / we can't reference something they already know. That means 5m saving/contributor.

So with 32h investe dwe're saving 20.8h over a 2 year period, a net loss of 11.2h.

That's not the math I'd do personally- where I suspect the change costs are higher and the time delta is lower (mainly because not all contributors are actually impacted).
